### PR TITLE
feat: Embeded PostgreSQL with mockdb data for self-hosted

### PR DESCRIPTION
**Pull request by sum35h**
## Description
This PR includes changes for self-hosted instances to utilize an embedded postgres database for the Template's mockdb.
After this change, by default Appsmith will always run the embedded Postgres, and connect to it instead of the mockdb cloud instance.

**Solves:** Issue[#20107](https://github.com/appsmithorg/appsmith/issues/20107)

**Changes:**
- [x]  Install Postgresql in the Docker container using Dockerfile  :- _Updated Docker file to install PostgreSQL v13_

- [x]  Add provision to stop/disable Postgresql from booting based on an environment variable :- _To disable embedded postgres, set the **ENV** var `APPSMITH_ENABLE_EMBEDDED_DB` to `0`;  Configured supervisord to control the lifecycle of the postgres server  ( Supervisor uses user postgres to start the server since root is not allowed)_

- [x] Seed data into the Postgresql DB during first container startup. This ensures that we don't have to ship the Docker 
      container with data & bloat it unnecessarily. :- _Curretly, the SQL dump is baked into the image._
      
- [x] Persist the Postgresql data in /appsmith-stacks. This ensures that the data is persisted across version upgrades. :- _The postgres uses the file system at` /stacks/data/postgres/main` ; note: The file system owner is user/group postgres_

- [x] Update the mock db endpoint in the product to use the local DB by default instead of the hosted DB for self-hosted instances. The cloud instance should still use the hosted DB :- _The embedded postgres authentication is set to the type `trust`, allowing the existing template mockdb meta data to be used without any update to the existing templates mockdb endpoint and credentials_

**Note: There is an additional table `mydb` added to the mockdb for quickly verifying the db source is the embedded postgres, since the cloud mockdb does not have that table.**

### CODE REVIEW

**Comment by pratapaprasanna** [comment]
Deploy-preview failing @sum35h

**Comment by sum35h** [comment]
Update: The embedded postgres does not work on AWS EFS network file system due to permission conflict.
`+ init_postgres
+ [[ '' != 1 ]]
+ echo '\nChecking initialized local postgres'
\nChecking initialized local postgres
+ POSTGRES_DB_PATH=/appsmith-stacks/data/postgres/main
+ '[' -e /appsmith-stacks/data/postgres/main/PG_VERSION ']'
+ echo 'Initializing local postgresql database'
`

**Comment by sum35h** [comment]
Issue with volume mounts:
Postgres does not allow it's server to be run with super user access, so postgres user is being used instead.
Postgres also needs it's filesystem to be owned by the same user who ran the server.

In case of EBS or EFS volumes, only root user has access to create directories.
The only way to change the ownership, seems to be to change the ownership of the whole volume.

File: deploy/docker/entrypoint.sh [hunk 1]
```diff
@@ -298,6 +304,57 @@ check_redis_compatible_page_size() {
   fi
 }
 
+init_postgres() {
+  # Initialize embedded postgres by default; set APPSMITH_DISABLE_EMBEDDED_POSTGRES to 1, to use existing cloud postgres mockdb instance
+  if [[ ${APPSMITH_DISABLE_EMBEDDED_POSTGRES: -0} != 1 ]]; then
```
[the remark this answers is not in the document]
Comment (sum35h) [reply deploy/docker/entrypoint.sh:309]: @mohanarpit Would this variable name be more confusing to the user if the user assumes it means the MongoDB?

[the remark this answers is not in the document]
Comment (sum35h) [reply deploy/docker/entrypoint.sh]: So, I understand later we will also move the mock data for Mongo into the embedded MongoDB and use the same ENV var for enabling it.

**Comment by sum35h** [comment]
I've tested this image on EKS and ECS, and works with EFS.
Tested it on ACI but fails with Azure File Storage
```
Initializing local postgresql database
+ echo 'Initializing local postgresql database'
+ mkdir -p /appsmith-stacks/data/postgres/main
+ chown postgres:postgres /appsmith-stacks/data/postgres/main
+ su -m postgres -c '/usr/lib/postgresql/13/bin/initdb -D /appsmith-stacks/data/postgres/main'
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "C.UTF-8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

fixing permissions on existing directory /appsmith-stacks/data/postgres/main ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 20
selecting default shared_buffers ... 400kB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
2023-03-06 13:41:38.889 UTC [1638] FATAL:  data directory "/appsmith-stacks/data/postgres/main" has wrong ownership
2023-03-06 13:41:38.889 UTC [1638] HINT:  The server must be started by the user that owns the data directory.
child process exited with exit code 1
initdb: removing contents of data directory "/appsmith-stacks/data/postgres/main"
+ su postgres -c '/usr/lib/postgresql/13/bin/pg_ctl -D /appsmith-stacks/data/postgres/main start'
pg_ctl: directory "/appsmith-stacks/data/postgres/main" is not a database cluster directory
+ seed_embedded_postgres
+ tar -xf /opt/appsmith/templates/postgres_sql.tar.gz -C /opt/appsmith/templates
+ psql -U postgres -c 'CREATE DATABASE mockdb;'
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ su postgres -c '/usr/lib/postgresql/13/bin/createuser mockdb -s'
createuser: error: could not connect to database template1: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ psql -U postgres -d mockdb --file=/opt/appsmith/templates/mockdb_postgres.sql
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ psql -U postgres -c 'CREATE DATABASE users;'
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ su postgres -c '/usr/lib/postgresql/13/bin/createuser users -s'
createuser: error: could not connect to database template1: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ psql -U postgres -d users --file=/opt/appsmith/templates/users_postgres.sql
psql: error: connection to server on socket "/var/run/postgresql/.s.PGSQL.5432" failed: No such file or directory
	Is the server running locally and accepting connections on that socket?
+ rm /opt/appsmith/templates/mockdb_postgres.sql /opt/appsmith/templates/users_postgres.sql
+ su postgres -c '/usr/lib/postgresql/13/bin/pg_ctl stop -D /appsmith-stacks/data/postgres/main'
pg_ctl: directory "/appsmith-stacks/data/postgres/main" is not a database cluster directory
+ runEmbeddedPostgres=0

```

**Review by sharat87 — APPROVED** [review]

**Pull request merged**